src/meltano/core/project_plugins_service.py

```diff
@@ -234,29 +236,33 @@ def find_plugin(
                 f"ignoring `@{profile_name}` in plugin name.",
             )
 
-        try:
-            plugin = next(
-                plugin
-                for plugin in self.plugins(ensure_parent=False)
-                if (
-                    plugin.name == plugin_name  # noqa: WPS222 (with too much logic)
-                    and (plugin_type is None or plugin.type == plugin_type)
-                    and (
-                        invokable is None
-                        or self.ensure_parent(plugin).is_invokable() == invokable
-                    )
-                    and (
-                        configurable is None
-                        or self.ensure_parent(plugin).is_configurable() == configurable
-                    )
+        for plugin in self.plugins(ensure_parent=False):
+            if (
+                plugin.name == plugin_name  # noqa: WPS222 (with too much logic)
+                and (plugin_type is None or plugin.type == plugin_type)
+                and (
+                    invokable is None
+                    or self.ensure_parent(plugin).is_invokable() == invokable
                 )
-            )
-
-            return self.ensure_parent(plugin)
-        except StopIteration as stop:
-            raise PluginNotFoundError(
-                PluginRef(plugin_type, plugin_name) if plugin_type else plugin_name,
-            ) from stop
+                and (
+                    configurable is None
+                    or self.ensure_parent(plugin).is_configurable() == configurable
+                )
+            ):
+                return self.ensure_parent(plugin)
+            elif plugin.type == PluginType.MAPPERS:
+                mapping_name = plugin.extra_config.get("_mapping_name")
+                if mapping_name == plugin_name:
+                    all_mappings = self.find_plugins_by_mapping_name(mapping_name)
+                    if len(all_mappings) > 1:
+                        raise click.ClickException(
+                            f"Ambiguous mapping name {mapping_name}, "
+                            "found multiple matches.",
+                        )
+                    return self.ensure_parent(plugin)
+        raise PluginNotFoundError(
+            PluginRef(plugin_type, plugin_name) if plugin_type else plugin_name,
+        )
```

Comment:
You can easily refactor this to something like:

```suggestion
                mapping = self._find_mapping(plugin_name, plugin)
                if mapping:
                    return mapping
        raise PluginNotFoundError(
            PluginRef(plugin_type, plugin_name) if plugin_type else plugin_name,
        )

    def _find_mapping(self, plugin_name: str, plugin: ProjectPlugin):
        mapping_name = plugin.extra_config.get("_mapping_name")
        if mapping_name == plugin_name:
            all_mappings = self.find_plugins_by_mapping_name(mapping_name)
            if len(all_mappings) > 1:
                raise click.ClickException(
                    f"Ambiguous mapping name {mapping_name}, "
                    "found multiple matches.",
                )
            return self.ensure_parent(plugin)
```

Although it's better not to use `click.ClickException` but a custom exception that inherits from `meltano.core.error.MeltanoError`

Reply:
@edgarrmondragon cool thanks, I'll get it updated. I was keeping the ClickException only because of the changes I had to make in the parser. Altering it was triggering other tests to fail but now that I've got everything working I can probably work through that easier.